pkg/codegen/hcl2/model/type.go

```diff
@@ -35,7 +35,7 @@ func (k ConversionKind) Exists() bool {
 type Type interface {
 	Definition
 
-	Equals(other Type) bool
+	Equals(other Type, seen map[Type]struct{}) bool
```

Comment:
Do you think that we can replace this with an unexported version of the same s.t. the public API remains identical?

```suggestion
	Equals(other Type) bool
	equals(other Type, seen map[Type]struct{}) bool
```

Also, is referential identity okay for the set, given that types with different references can have the same identity?

Reply:
> Do you think that we can replace this with an unexported version of the same s.t. the public API remains identical?

We can.  It unfortunately means a lot of boilerplate in every implementation (since they all need identical implementations of `Equals`.  But that's fine I suppose.  I wasn't sure what the real "public contract" was here as there are very few external callers of this (I only found one).

I'll make the change though to avoid any public API churn.

> Also, is referential identity okay for the set, given that types with different references can have the same identity?

The set watches the visited receiver types, but doesn't pay attention to the "other" types.  So the only risk here would be that somehow there was recursion in the receiver type graph that didn't have referential identity.  I honestly don't *think* that should be possible.   Unless I am misunderstanding the question?